distributed_shampoo/shampoo_fsdp_utils.py

```diff
@@ -0,0 +1,900 @@
+        use_dtensor: bool = False,
+        left_comm: CommunicationType = CommunicationType.NONE,
+        right_comm: CommunicationType = CommunicationType.NONE,
+    ):
+        super(CommunicationShampooPreconditioner, self).__init__(
+            param,
+        )
+
+        # Set parameters.
+        self._large_dim_method = large_dim_method
+        self._beta2 = beta2
+        self._epsilon = epsilon
+        self._exponent_override = exponent_override
+        self._exponent_multiplier = exponent_multiplier
+        self._use_bias_correction = use_bias_correction
+        self._max_preconditioner_dim = max_preconditioner_dim
+        self._dtype = dtype
+        self._idx = idx
+        self._start_preconditioning_step = start_preconditioning_step
+        self._use_merge_dims = use_merge_dims
+
+        _, orig_shape, orig_numels, shard_param_info = metadata
+        start_idx = shard_param_info.intra_param_start_idx
+        end_idx = shard_param_info.intra_param_end_idx
+        self._orig_shape = orig_shape
+        self._start_idx = start_idx
+        self._end_idx = end_idx
+
+        self._left_comm = left_comm
+        self._right_comm = right_comm
+
+        # Initialize exponential moving average
+        self.bias_correction1 = torch.as_tensor(1.0)
+        self.exp_avg = torch.zeros_like(param, memory_format=torch.preserve_format)
+
+        # Initialize communication buffers
+        # TODO: decide whether to keep checks on Nones or update left_comm and right_comm
+        self.left_send_buffer = None
+        self.right_send_buffer = None
+        self.left_recv_buffer = None
+        self.right_recv_buffer = None
+        # TODO: is there a cleaner way of keeping track of ops than these lists? len of each list is at most 2
+        self.forward_ops = []
+        self.backward_ops = []
+        split_param = convex_split(
+            param, orig_shape, start_idx, end_idx, include_placeholder=True
+        )
+
+        if (
+            left_comm == CommunicationType.SEND
+            # TODO: these checks could be moved into the main optimizer, but that may get messy
+            # reconsider if a more complex assignment function is written
+            and start_idx != 0
+            and split_param[0].numel() != 0
+        ):
+            self.left_send_buffer = torch.zeros_like(split_param[0]).to(param.device)
+            # TODO: not sure if calling dist.get_rank() is good practice; may need to pass rank into the class
+            self.forward_ops.append(
+                dist.P2POp(
+                    dist.isend, self.left_send_buffer, dist.get_rank() - 1, tag=idx
+                )
+            )
+            self.backward_ops.append(
+                dist.P2POp(
+                    dist.irecv,
+                    self.left_send_buffer,
+                    dist.get_rank() - 1,
+                    tag=idx,
+                )
+            )
+            split_param = split_param[1:]
```

Comment:
This makes a copy. Can we use pop instead?
```suggestion
            split_param.pop(0)
```